### ISSUE TITLE
Bug Issue: Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

### ISSUE THREAD

**Issue by BabyWaffles** [ultralytics/ultralytics#9833]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Got the "Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions." when trying to train a "v8" "n" model with more than 10 epochs

### Environment

Ultralytics YOLOv8.1.23 🚀 Python-3.11.9 torch-2.1.2+cu118 CUDA:0 (Tesla V100, 15102MiB)
Setup complete ✅ (16 CPUs, 31.9 GB RAM, 1959.9/3514.1 GB disk)

OS                  Windows 11 Pro 10.0.22621 Build 22621
Environment         venv
Python              3.11.8
Install             git
RAM                 32.00 GB
CPU                 AMD Ryzen 7 3700x
CUDA                True

### Minimal Reproducible Example

from ultralytics import YOLO
import torch
from os import getcwd

def main():
  #Checks torch existence and version
  print(torch.__version__)

  #Define and Declare the working directory path
  working_path = getcwd()

  #Define the models directory path
  models_path = working_path + '\\models'
  models_path_v8 = models_path + '\\v8'
  models_path_v9 = models_path + '\\v9'
  
  #Model variables

  #yolov8
  large = models_path_v8 + '\\yolov8l.pt' # YoloV8 Large Model
  medium = models_path_v8 + '\\yolov8m.pt' # YoloV8 Medium Model
  nano = models_path_v8 + '\\yolov8n.pt' # YoloV8 Nano Model
  small = models_path_v8 + '\\yolov8s.pt' # YoloV8 Small Model
  extreme_large = models_path_v8 + '\\yolov8x.pt' # YoloV8 Extremely Large Model

  #yolov9
  extreme_large = models_path_v9 + '\\yolov9e.pt' # YoloV9 Extremely Large Model

  #Type Declared Fixed YAML Data Path
  yaml_data = working_path + '\\person-data.yaml' # Yaml Data Path File to Define the directory of the training data and it's structure

  #Save training sessions
  path_sessions = working_path + '.\\runs\\detect\\person-sample' # Directory path to store the training session output

  #Training Portion
  model = YOLO(nano)  # Train and transfer weights from a predefined model
  model.train(data = yaml_data , epochs=150 , pretrained=False , device=0, amp=True,  optimizer='AdamW', patience=20 ,project=path_sessions) 

if __name__ == "__main__": # Defines main and start the function (main)
  main()

#End of Script

### Additional

If the reproducible sample code looks weird, please use the RAW version due to GitHub's text formatting.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by BabyWaffles** [comment]
Here is the person-data.yaml file contents if you are wondering.

path: E:\.intRepository\person_yolov8\training_data\person_generated
train: E:\.intRepository\person_yolov8\training_data\person_generated\train
val: E:\.intRepository\person_yolov8\training_data\person_generated\Val
#test: 'training_data\person_generated\test'

nc: 1
names: 
- 0: person

**Comment by BabyWaffles** [comment]
the 0:person portion has a"-" instead. Ignore the "." generated by GitHub due to text formatting.

**Comment by glenn-jocher** [comment]
@BabyWaffles hello! Thanks for sharing the details. It looks like there's a slight formatting issue in your `person-data.yaml` for the `names` field. It should be a list without the numbering for each item. Here's a quick fix:

```yaml
nc: 1
names: 
  - person
```

Just replace the `names` part in your current YAML file with the snippet above, and you should be all set! Let me know if that does the trick for you. Happy training! 🚀

**Comment by BabyWaffles** [comment]
Nope. Replaced it and still got Torch Use Cuda Device Side Assertion triggered.

**Comment by glenn-jocher** [comment]
Hey @BabyWaffles 👋, thanks for giving that a try! The device-side assertion issue often hints at a problem related to the CUDA environment or an operation that the model is attempting to perform on the GPU which it can't handle as expected.

Could you please try setting `amp=False` to disable automatic mixed precision to see if it's related to AMP? Like this:

```python
model.train(data=yaml_data, epochs=150, pretrained=False, device=0, amp=False, optimizer='AdamW', patience=20, project=path_sessions)
```

It's a shot in the dark, but sometimes AMP can introduce complexities, especially with specific CUDA versions.

Let's see how that goes! 🤞

**Comment by BabyWaffles** [comment]
I've located the bug. It was my training data index problem. I used the one that has more than 1 index so it triggered the DSA as the labelled data is on index, 1 with the following coordinates.

So basically e.g., the coordinate of the labelled data is at xyz, but the index was set to 1 instead of 0.

Incorrect example: 1 0.0027771, 0.0027251, 0.002661
Correct example: 0 0.0027771, 0.0027251, 0.002661

**Comment by glenn-jocher** [comment]
Hey @BabyWaffles 👋, glad to hear you found the source of the issue! Index errors in label files can indeed cause unexpected behavior. Thanks for suggesting an enhancement for better handling and clearer messaging on such errors. We'll definitely consider adding a more descriptive check for label indexes to assist users in troubleshooting similar situations in the future. Your input is highly valued, and great job debugging your training setup! If there's anything else you need or further ideas you have, feel free to share. Happy training! 😊
